### PR TITLE
GH-2640: tensor forward

### PR DESCRIPTION
**Pull request by helpmefindaname**
First PR about https://github.com/flairNLP/flair/issues/2640: refactoring models, such that each model has a `_prepare_tensors` and a `forward` method, where one extracts all tensors out of the data points and the latter only does tensor computations.

That way, JIT tracing and ONNX conversation of the models should be possible.
**Notice:** this is not expected to give a huge speedup, as most computation is within the embeddings, which won't be affected. Due to that, the ONNX conversion is also not documented. 

This PR also adds unit scaling for model downloads when a non-huggingface model is downloaded.
This PR also fixes a bug, that relation extraction models without weight_dict set can be loaded.
This PR also fixes an encoding error, if glue-mnli is loaded on a windows machine
This PR also adds an option to not add a unk token to labels.

### CODE REVIEW

**Comment by helpmefindaname** [comment]
Already working is the TextClassifier, see the following example code:
```
import torch
from flair.data import Sentence
from flair.models import TextClassifier


def convert():
    model = TextClassifier.load("de-offensive-language")
    example_sentence = Sentence("This is a sentence.")

    tensors = model._prepare_tensors([example_sentence])

    torch.onnx.export(model, tensors, "texclassifier.onnx", input_names=["text_embedding_tensor"], output_names=["scores"], opset_version=7, verbose=True)


if __name__ == "__main__":
    convert()
```

**Comment by bratao** [comment]
Any chance of this making into main tree? I love flair, and ONNX models are a must for production code!

**Comment by alanakbik** [comment]
hello @bratao yes we are looking into this now - sorry @helpmefindaname for taking so long, I first wanted to release Flair 0.11 (done yesterday) before making bigger changes.

File: flair/nn/model.py [hunk 1]
```diff
@@ -582,42 +613,19 @@ def multi_label_threshold(self, x):  # setter method
         else:
             self._multi_label_threshold = {"default": x}
 
-    def forward_loss(self, sentences: Union[List[DT], DT]) -> Tuple[torch.Tensor, int]:
-
-        # make a forward pass to produce embedded data points and labels
-        embedded_data_points, labels = self.forward_pass(sentences)  # type: ignore
-
-        # no loss can be calculated if there are no labels
-        if not any(labels):
```
Comment (alanakbik) [inline flair/nn/model.py:591]: This check is currently important for models that require candidates like the EntityLinker or RelationExtractor.

**Comment by alanakbik** [comment]
@helpmefindaname I started reviewing, but it will take some time to think through all the changes. 

The logic of splitting out the tensor and non-tensor stuff in the forward pass makes sense for ONNX but I worry about code readability with the logic now distributed across many methods and different parent classes (`forward_loss` now calls `forward` which in turn calls `forward_pass`) and the risk of tensor and label preparation diverging. Before with the `forward_pass` there was a single method that encapsulates all model-specific logic for each class that inherits from DefaultClassifier, albeit with the drawback of an overly complex return type. I wonder if the structure can be somehow more simplified but have to think about this for a bit. 

There also is a small problem with models that require candidates and get passed sentences without candidates, see:

```python
# init a random linker for testing
linker: EntityLinker = EntityLinker(TransformerWordEmbeddings(model='distilbert-base-uncased'),
                                    label_dictionary=Dictionary())

# sentence with candidate label - works
sentence = Sentence("I live in Berlin")
sentence[3:4].add_label('nel', 'LOC')
linker.predict(sentence)
print(sentence)

# sentence without candidate - fails
sentence = Sentence("I live in Berlin")
linker.predict(sentence) 
```

**Comment by alanakbik** [comment]
Hello @helpmefindaname from a first look-through I like this structure a lot!

Some suggestions: 
- `forward_pass` could be renamed into something that communicates that this is now only _optional_ intermediary layers between the embeddings and the decoder.  (Since it always takes an embedding Tensor as input and the output goes to the decoder.) Maybe something that does not contain "forward" to avoid confusion with "forward_loss" and "forward". Unfortunately, I cannot come up with a good name now... 
- `DefaultClassifier` could have a standard implementation of  `_get_label_of_datapoint` (i.e. make it non-abstract). In most (all?) cases `return [datapoint.get_label(self._label_type).value]` should already do the trick, I think. That would be one less method to implement by default. 

That would reduce the code needed to 6 methods for a standard default classifier with no extras (for instance the `TextPairClassifier`): 

1. an `__init__`
2. a `_get_prediction_data_points` 
3. a `_embed_prediction_data_point` 
4. a `label_type` property
5. & 6. and the two load/save methods `_get_state_dict` and `_init_model_with_state_dict`

with all actual important logic in the first three methods, which is cool.

**Comment by helpmefindaname** [comment]
A small change to the interface:
Instead of providing the embedding as a parameter, I had to create a property for it, to not double the amount of parameters per model. Also the embedding is optional so `TextPairClassifier` can choose to not have pre-embedded sentences but do the embedding on its own.

The RelationExtractor logic could be simplified, as the logic for extracting labels was the same as the logic created in `_PartOfSentence` thus `Relation(span1, span2).get_label(self.label_type)` was sufficient.

File: flair/nn/model.py [hunk 2]
```diff
@@ -628,7 +673,24 @@ def _calculate_loss(self, scores, labels) -> Tuple[torch.Tensor, int]:
...
+
+        # make a forward pass to produce embedded data points and labels
+        predict_data_points = self._get_prediction_data_points(sentences)
+        labels = self._prepare_label_tensor(predict_data_points)
+
+        # no loss can be calculated if there are no labels
+        if labels.size(0) == 0:
+            return torch.tensor(0.0, requires_grad=True, device=flair.device), 1
+
+        embedded_tensor = self._prepare_tensors(sentences)
```
Comment (alanakbik) [inline flair/nn/model.py:686]: This could cause some problems I think. 4 lines above the `predict_data_points` are extracted using `self._get_prediction_data_points(sentences)` and labels extracted from them. 

But here now the `sentences` are passed into `self._prepare_tensors` which first applies a filter, then again calls `self._get_prediction_data_points(sentences)`. So it is possible that the datapoints extracted after the filter diverge from the datapoints used to get the labels. 

Maybe this `_prepare_tensors` could take as input `predict_data_points` instead of `sentences`? The filtering could be done beforehand as first line in `forward_loss`.

File: flair/models/entity_linker_model.py [hunk 3]
```diff
@@ -62,75 +63,35 @@ def __init__(
...
+        return entities
 
-        if for_prediction:
-            return embedded_entity_pairs, span_labels, data_points
+    def _filter_data_point(self, data_point: Sentence) -> bool:
+        return bool(data_point.get_labels(self.label_type))
 
-        return embedded_entity_pairs, span_labels
+    def _embed_prediction_data_point(self, prediction_data_point: Span) -> torch.Tensor:
+        return self.aggregated_embedding(prediction_data_point, self.word_embeddings.get_names()).unsqueeze(0)
```
[the remark this answers is not in the document]
Comment (helpmefindaname) [reply flair/models/entity_linker_model.py]: fixed and added a test

File: flair/nn/model.py (hunk 2 above)
Comment (helpmefindaname) [reply flair/nn/model.py:688]: Very good catch!
Although the filtering is usually for sentences that do not have any labels, it could be misused and create some divergence.

I moved the filtering to be the first thing to be done to a batch. Even before calling `_prepare_tensors`.

Sadly, I cannot change the signature of `_prepare_tensors`, as otherwise it wouldn't be in line with the general `flair.nn.Model` and would make it very complicated to use jit or onnx exports.

**Comment by alanakbik** [comment]
@helpmefindaname thanks for fixing the rebase conflicts!

**Review by alanakbik — APPROVED** [review]
Thanks a lot for adding this to Flair!
